drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -35,1403 +35,1532 @@
-							int w = current_clip->final_clip_rect.size.x;
-							int h = current_clip->final_clip_rect.size.y;
+										glScissor(x, y, w, h);
 
-							if (storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_VFLIP])
-								y = current_clip->final_clip_rect.position.y;
+										reclip = false;
+									}
+								}
+							}
 
-							glScissor(x, y, w, h);
+						} break;
 
-							reclip = false;
-						}
+						default: {
+							// FIXME: Proper error handling if relevant
+							//print_line("other");
+						} break;
 					}
 				}
 
-			} break;
-
-			default: {
-				// FIXME: Proper error handling if relevant
-				//print_line("other");
-			} break;
+			} // default
+			break;
 		}
 	}
+
+	// zero all the batch data ready for a new run
+	bdata.reset_flush();
 }
 
-void RasterizerCanvasGLES2::_copy_screen(const Rect2 &p_rect) {
+void RasterizerCanvasGLES2::render_joined_item_commands(const BItemJoined &bij, Item *current_clip, bool &reclip, RasterizerStorageGLES2::Material *p_material) {
 
-	if (storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_DIRECT_TO_SCREEN]) {
-		ERR_PRINT_ONCE("Cannot use screen texture copying in render target set to render direct to screen.");
-		return;
+	Item *item = 0;
+
+	FillState fill_state;
+	fill_state.reset();
+	fill_state.use_hardware_transform = bij.use_hardware_transform();
+
+	for (unsigned int i = 0; i < bij.num_item_refs; i++) {
+		item = bdata.item_refs[bij.first_item_ref + i].m_pItem;
+
+		int command_count = item->commands.size();
+		int command_start = 0;
+
+		while (command_start < command_count) {
+			// fill as many batches as possible (until all done, or the vertex buffer is full)
+			bool bFull = prefill_joined_item(fill_state, command_start, item, current_clip, reclip, p_material);
+
+			if (bFull) {
+				// flush
+				flush_render_batches(item, current_clip, reclip, p_material);
+				fill_state.reset();
+			}
+		}
 	}
 
-	ERR_FAIL_COND_MSG(storage->frame.current_rt->copy_screen_effect.color == 0, "Can't use screen texture copying in a render target configured without copy buffers.");
+	// flush if any left
+	if (item) {
+		// always pass first item
+		item = bdata.item_refs[bij.first_item_ref].m_pItem;
 
-	glDisable(GL_BLEND);
+		flush_render_batches(item, current_clip, reclip, p_material);
+	}
+}
 
-	Vector2 wh(storage->frame.current_rt->width, storage->frame.current_rt->height);
+void RasterizerCanvasGLES2::flush_render_batches(Item *p_item, Item *current_clip, bool &reclip, RasterizerStorageGLES2::Material *p_material) {
+	//print_line("_flush_render_batches");
 
-	Color copy_section(p_rect.position.x / wh.x, p_rect.position.y / wh.y, p_rect.size.x / wh.x, p_rect.size.y / wh.y);
+	// some heuristic to decide whether to use colored verts.
+	// feel free to tweak this.
+	// this could use hysteresis, to prevent jumping between methods
+	// .. however probably not necessary
+	if ((bdata.total_color_changes * 4) > (bdata.total_quads * 3)) {
```

Comment:
I wonder if it would be worth studying this. Could set a ratio in project settings and then test.

Reply:
Yes, it would be possible to test further, this cutoff was pure guesswork. However it is possible the best value may depend on the hardware, larger vertex format versus number of drawcalls. It also may be relatively more important to keep the vertex format small for large batches (and vice versa). Keeping it simple to start with.

Reply:
Should probably make it a project setting then.